### DISCUSSION TITLE
Computed number field for data model relations

### DISCUSSION THREAD

**Discussion by Taldres** [twentyhq/twenty#7020] [Ideas]
## Scope & Context

It would be really cool to have a kind of calculated sum in the data models, which aggregates the values of a specific field from all entries of a model that are related to this one.

Like the number of machines from multiple production plants that belong to the model.

## Current behavior

I added a `Machines` number field to the `Company` model, which defines the total number of machines within a company. To get more detailed and structured information, I created a `Production Site` model, which belongs to a `Company` and also has a `Machines` number field.

Currently, I have to set the total number of machines manually.

<img width="943" alt="Bildschirmfoto 2024-08-31 um 12 48 37" src="https://github.com/user-attachments/assets/f86c6eb2-7de6-48cb-bddb-ddbb27d316a0">

<img width="474" alt="Bildschirmfoto 2024-08-31 um 12 26 10" src="https://github.com/user-attachments/assets/a04dbad5-8fc3-42e8-b52b-6e52588eaaac">
<img width="474" alt="Bildschirmfoto 2024-08-31 um 12 26 15" src="https://github.com/user-attachments/assets/b9e0977e-de5a-4a5d-a9a1-cb36e493dd8d">

## Expected behavior
A `computed number` or `sum number` field that I could link to my related model to calculate the expected total amount.

<img width="825" alt="Bildschirmfoto 2024-08-31 um 12 25 36" src="https://github.com/user-attachments/assets/7785c00b-7a49-4370-82e4-a20ea2297def">

**Comment by Sandheep-OSC** [comment]
This looks interesting to me. I would like to work in it.

**Comment by charlesBochet** [comment]
Hi @Taldres, this is indeed something we have in mind.

The product still needs to be refined about it.
This is touching two areas we are still refining: Computed Fields and aggregations.

We will introduce a new type of field we would call a "Computed Field" (or a "LookUp Field"). The tricky part is that we cannot compute this at read without introducing major performance issues on index views ; this would indeed force the engine to join additional tables to get the data and can slow down the app very quickly. Instead we would probably want to compute these Computed Field at write time and would probably leverage the workflow/automation work that is ongoing

Aggregations: regarding your specific example, it also reminds me of aggregations. We will likely introduce new rest / graphql endpoints to get basic aggregates on tables (sum, average, ...) and add the possibility to query them while querying relations. Here we could for example allow to display the number of "machines" next to the production sites

I think it's still a bit early for Twenty to tackle this but it's definitely something we have in mind

**Comment by Taldres** [comment]
> Hi @Taldres, this is indeed something we have in mind.
> 
> The product still needs to be refined about it. This is touching two areas we are still refining: Computed Fields and aggregations.
> 
> We will introduce a new type of field we would call a "Computed Field" (or a "LookUp Field"). The tricky part is that we cannot compute this at read without introducing major performance issues on index views ; this would indeed force the engine to join additional tables to get the data and can slow down the app very quickly. Instead we would probably want to compute these Computed Field at write time and would probably leverage the workflow/automation work that is ongoing
> 
> Aggregations: regarding your specific example, it also reminds me of aggregations. We will likely introduce new rest / graphql endpoints to get basic aggregates on tables (sum, average, ...) and add the possibility to query them while querying relations. Here we could for example allow to display the number of "machines" next to the production sites
> 
> I think it's still a bit early for Twenty to tackle this but it's definitely something we have in mind

Thank you for your detailed response. I'm glad to hear that you already have this in mind. I'm looking forward to the future solution.

**Comment by FelixMalfait** [comment]
Thanks for the suggestion! Converting this to a discussion as this is planned on the roadmap. I hope we'll have something powerful by end of Q1 next year.

**Comment by jgoyvaerts** [comment]
Any update on this? This is blocking me from trying out Twenty. Or is there a workaround possible for now?

**Reply by AltoresMonaco** [reply]
custom automation through webhooks i guess ! using unique ids on fields and updating data through the api

**Comment by pmckeown** [comment]
I was hoping to try this computed field capability using the workflow feature, but its not well documented.  Are there any guides or snippets that show how to reference the Opportunities record that is being updated in this example?

Also - are workflows only going to be a paid-for feature?

<img width="340" alt="image" src="https://github.com/user-attachments/assets/bdc0cf95-4918-495e-90c2-bd10718b666c" />

<img width="397" alt="image" src="https://github.com/user-attachments/assets/5515f324-9200-45d1-bc09-0ac262e66af0" />

**Reply by FelixMalfait** [reply]
Hey, there will be a cap on the number of runs on the cloud version but it's going to be generous. 

I'm sorry it's not well documented yet. We still need to add some features before we deem it "production ready" (branch/conditions, loops, delay, etc.)

**Comment by BOHEUS** [comment]
Related to https://github.com/twentyhq/core-team-issues/issues/813
